src/app/content/courses/create-your-sdk-with-codama/codama-from-scratch/en.mdx

```diff
@@ -0,0 +1,230 @@
+
+```ts
+const pdaSeedNode = variablePdaSeedNode('authority', publicKeyTypeNode())
+```
+
+Here's how they're used in a `pdaNode`:
+
+```ts
+const constantPda = pdaNode({
+    name: 'tickets',
+    seeds: [constantPdaSeedNodeFromString('utf8', 'tickets')],
+}); 
+
+const variablePda = pdaNode({
+    name: 'ticket',
+    seeds: [variablePdaSeedNode('authority', publicKeyTypeNode())],
+});
+```
+
+> `PdaSeedNode` is a type alias and cannot be used directly as a node. When a `PdaSeedNode` is required, use one of the specific pda seed node types instead.
+
+<ArticleSection name="Writing a Codama IDL" id="writing-a-codama-idl" level="h2" />
+
+Now that we have examined the most important nodes available in a program, let's discover how to create a Codama IDL from scratch.
+
+### Root Node
+
+To create the foundation of your Codama IDL, we use `RootNode`. This node serves as the top-level container that holds your main `ProgramNode` as well as any additional programs that may be referenced by the main program.
+
+```ts
+const node = rootNode(programNode({ ... }));
+```
+
+### Program Node
+
+To define an entire on-chain program, we use `ProgramNode`. This node represents the complete program deployed on-chain and defines all the minimum viable elements such as accounts, instructions, PDAs, and errors.
+
+In addition to these core elements, this node accepts the program's name, version, deployment public key, and markdown documentation:
+
+
+```ts
+const node = programNode({
+    name: 'counter',
+    publicKey: '22222222222222222222222222222222222222222222',
+    version: '0.0.1',
+    docs: [],
+    accounts: [],
+    instructions: [],
+    definedTypes: [],
+    pdas: [],
+    errors: [],
+});
+```
+
+In the next section we'll examine all these elements in detail.
+
+### Account Node
+
+To define on-chain accounts, we use `AccountNode`. This node is characterized by its name, data structure, and optional attributes such as PDA definitions and account discriminators. 
+
+It represents the accounts that typically correspond to the `state.rs` file of your program.
+
+The docs field can be used to add documentation explaining what this account accomplishes within the program:
+
+```ts
+const node = accountNode({
+    name: 'token',
+    data: structTypeNode([
+        structFieldTypeNode({ name: 'mint', type: publicKeyTypeNode() }),
+        structFieldTypeNode({ name: 'owner', type: publicKeyTypeNode() }),
+        structFieldTypeNode({ name: 'amount', type: numberTypeNode('u64') }),
+    ]),
+    discriminators: [sizeDiscriminatorNode(72)],
+    size: 72,
+    pdas: [
+        pdaNode({
+            name: 'associated_token_account',
+            seeds: [
+                constantPdaSeedNodeFromString('utf8', 'token'),
+                variablePdaSeedNode('mint', publicKeyTypeNode()),
+                variablePdaSeedNode('owner', publicKeyTypeNode()),
+            ],
+        }),
+    ],
```

Comment:
This is actually incorrect, the `AccountNode` does not store an array of `PdaNode` but a single optional link to a `PdaNode` via a `PdaLinkNode`.

This is so that, if a `PdaNode` defined inside a `ProgramNode` refers to the an on-chain account offered by that program, the two can be linked together.

```suggestion
    pda: pdaLinkNode("associatedTokenAccount"),
```